### PR TITLE
Remove metric aggregator Subtract interface

### LINKED ISSUE #2349: Remove the metric aggregator Subtractor interface
### Problem Statement

The Metrics SIG has ruled out specifying that SDKs support a cumulative-to-delta conversion.

### Proposed Solution

I advocated for this support not to be standardized because of the complexity it generates, as seen in the OTel-Go SDK. I previously hacked this removal into #1415 to show how much complexity can be removed.

#### Alternatives

The alternative is to keep this support, but we have determined that all systems we know of would prefer to convert cumulative temporality sums to gauge values than into delta temporality sums.

### PR DESCRIPTION
**Pull request by jmacd**
This PR removes a bunch of complexity that the Metrics SIG group decided not to specify for the default SDK. The implication is that you will no longer be able to convert asynchronous counters into delta counters for export, which is no great loss.

Fixes #2349.

### CODE REVIEW

File: sdk/metric/processor/basic/basic.go [hunk 1]
```diff
@@ -372,16 +350,23 @@ func (b *state) ForEach(exporter aggregation.TemporalitySelector, f func(export.
 		case aggregation.DeltaTemporality:
 			// Precomputed sums are a special case.
 			if mkind.PrecomputedSum() {
-				agg = value.delta.Aggregation()
-			} else {
-				agg = value.current.Aggregation()
+				// This functionality was removed from
+				// the basic processor in PR #xxxx.
```
Comment (cbandy) [inline sdk/metric/processor/basic/basic.go:354]: 📝 Should this be #2350 (this PR?)

File: sdk/metric/processor/basic/basic_test.go [hunk 2]
```diff
@@ -183,28 +185,27 @@ func testProcessor(
...
 
 				if mkind.Asynchronous() {
 					// Asynchronous tests accumulate results multiply by the
 					// number of Accumulators, unless LastValue aggregation.
 					// If a precomputed sum, we expect cumulative inputs.
 					if mkind.PrecomputedSum() {
-						if aggTemp == aggregation.DeltaTemporality && akind != aggregation.LastValueKind {
-							multiplier = int64(nAccum)
+						if aggTemp == aggregation.DeltaTemporality {
+							panic("Impossible")
```
[the remark this answers is not in the document]
Comment (MadVikingGod) [reply sdk/metric/processor/basic/basic_test.go:208]: This is part of a test.  I think it prevents the test writer from using that combination of aggregator and meter.

**Comment by MadVikingGod** [comment]
LGTM, needs attention to get lint to pass.

**Comment by evantorrie** [comment]
Looks like this needs a `make precommit` to ensure gofmt has run on everything?

File: sdk/metric/processor/basic/basic_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pirgeo) [reply sdk/metric/processor/basic/basic_test.go:208]: Oh, whoops, I missed that :sweat_smile: Nevermind then :grimacing:

**Review by bogdandrutu — APPROVED** [review]

File: sdk/metric/processor/basic/basic.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/processor/basic/basic.go:354]:
```suggestion
				// the basic processor in PR #2350.
```

File: sdk/metric/processor/basic/basic_test.go [hunk 3]
```diff
@@ -183,28 +185,27 @@ func testProcessor(
...
 						} else if akind == aggregation.LastValueKind {
```
Comment (MrAlias) [inline sdk/metric/processor/basic/basic_test.go:209]:
```suggestion
						require.NotEqual(t, aggTemp, aggregation.DeltaTemporality)
						if akind == aggregation.LastValueKind {
```

**Review by MrAlias — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -31,6 +31,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Metric SDK removes the "exact" aggregator for histogram instruments, as it performed a non-standard aggregation for OTLP export (creating repeated Gauge points) and worked its way into a number of confusing examples. (#2348)
 
+### Removed
+
+- Remove the metric Processor's ability to convert cumulative to delta aggregation temporality. (#2350)
+
```
Comment (MrAlias) [inline CHANGELOG.md:37]: Looks like the sync with main has put this in the last release. Please move this to the `Unreleased` section.

**Review by MrAlias — COMMENTED** [review]
Looks ready to merge once we can fix the changelog

File: CHANGELOG.md (hunk 4 above)
Comment (jmacd) [reply CHANGELOG.md]: Fixed

**Comment by MrAlias** [comment]
> Looks ready to merge once we can fix the changelog

Gahh, I spoke too soon. I forgot that @bogdandrutu is not in the @open-telemetry/go-approvers group. We need one more approval.

**Review by MadVikingGod — APPROVED** [review]

**Pull request merged**